### PR TITLE
Add type hints to fileStore files

### LINKED ISSUE #3704: Add type hints to abstractFileStore.py
Add type hints to `src/toil/abstractFileStore.py` so it can be checked under mypy during linting.

Refers to #3568.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-955)
┆Issue Number: TOIL-955

### PR DESCRIPTION
**Pull request by jonathanxu18**
Adds type hints to: 

* `src/toil/fileStores/__init__.py`
* `src/toil/abstractFileStore.py` 
*  `src/toil/fileStores/nonCachingFileStore.py`

Fixes #3704, #3711, #3712.

Refers to #3568.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * PR submitter writes their recommendation for a changelog entry here

## Reviewer Checklist

 * [x] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [x] Read through the code changes. Make sure that it doesn't have:
    * [x] Addition of trailing whitespace.
    * [x] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [x] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [x] New functions or classes without informative docstrings.
    * [x] Changes to semantics not reflected in the relevant docstrings.
    * [x] New or changed command line options for Toil workflows that are not reflected in `docs/running/cliOptions.rst`
    * [x] New features without tests.
* [x] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [x] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

File: src/toil/fileStores/abstractFileStore.py [hunk 1]
```diff
@@ -17,9 +17,9 @@
 from abc import ABC, abstractmethod
 from contextlib import contextmanager
 from threading import Event, Semaphore
-from typing import Callable, Generator, Union
+from typing import Any, Dict, List, Set, Callable, ContextManager, Generator, IO, Iterator, Optional, Tuple, Union, TYPE_CHECKING
 
-import dill
+import dill # type: ignore
```
Comment (mr-c) [inline src/toil/fileStores/abstractFileStore.py:22]: Please add any needed stubs to `contrib/typeshed`.

File: src/toil/fileStores/abstractFileStore.py [hunk 2]
```diff
@@ -158,36 +161,33 @@ def open(self, job: Job) -> Generator[None, None, None]:
...
         Get a new local temporary directory in which to write files that persist for the duration of
         the job.
 
         :return: The absolute path to a new local temporary directory. This directory will exist
                  for the duration of the job only, and is guaranteed to be deleted once the job
                  terminates, removing all files it contains recursively.
-        :rtype: str
         """
-        return os.path.abspath(tempfile.mkdtemp(dir=self.localTempDir))
+        return os.path.abspath(tempfile.mkdtemp(prefix="t", dir=self.localTempDir))
```
Comment (mr-c) [inline src/toil/fileStores/abstractFileStore.py:173]: Why this change?

**Review by mr-c — CHANGES_REQUESTED** [review]

File: src/toil/fileStores/abstractFileStore.py (hunk 2 above)
Comment (jonathanxu18) [reply src/toil/fileStores/abstractFileStore.py]: My bad, I think I accidentally included that. Just removed it.

File: contrib/typeshed/dill/_dill.pyi [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from typing import Any, Optional
+
+def dump(obj: Any, file: Any, protocol: Optional[Any] = ..., byref: Optional[Any] = ..., fmode: Optional[Any] = ..., recurse: Optional[Any] = ..., **kwds: Any) -> None: ...
+def dumps(obj: Any, protocol: Optional[Any] = ..., byref: Optional[Any] = ..., fmode: Optional[Any] = ..., recurse: Optional[Any] = ..., **kwds: Any): ...
+def load(file: Any, ignore: Optional[Any] = ..., **kwds: Any): ...
+def loads(str: Any, ignore: Optional[Any] = ..., **kwds: Any): ...
```
Comment (mr-c) [inline contrib/typeshed/dill/_dill.pyi:6]: Are all these `Any`s really needed? More specific type hints would be appreciated :-)

File: src/toil/fileStores/abstractFileStore.py [hunk 4]
```diff
@@ -77,37 +80,37 @@ def __init__(self, jobStore: AbstractJobStore, jobDesc: JobDescription, localTem
...
+        self.filesToDelete: Set[str] = set()
         # Records IDs of jobs that need to be deleted when the currently
         # running job is cleaned up.
         # May be modified by the worker to actually delete jobs!
-        self.jobsToDelete = set()
+        self.jobsToDelete: Set[str] = set()
         # Holds records of file ID, or file ID and local path, for reporting
         # the accessed files of failed jobs.
-        self._accessLog = []
+        self._accessLog: List[Any] = []
```
Comment (mr-c) [inline src/toil/fileStores/abstractFileStore.py:102]: Can you be more specific here? Hint, this will be my response to all `Any`s :-)

File: src/toil/fileStores/abstractFileStore.py [hunk 5]
```diff
@@ -430,14 +432,14 @@ class _StateFile(object):
         """
         Utility class to read and write dill-ed state dictionaries from/to a file into a namespace.
         """
-        def __init__(self, stateDict):
+        def __init__(self, stateDict: Dict[Any, Any]):
```
Comment (mr-c) [inline src/toil/fileStores/abstractFileStore.py:435]: Not even `Dict[str, Any]`?

File: src/toil/fileStores/abstractFileStore.py [hunk 6]
```diff
@@ -447,11 +449,12 @@ def open(cls, outer=None):
             raise NotImplementedError()
 
         @classmethod
-        def _load(cls, fileName):
+        def _load(cls, fileName: str) -> Any:
```
Comment (jonathanxu18) [inline src/toil/fileStores/abstractFileStore.py:452]: I tried returning `'_StateFile'` here instead but it kept saying that it is not defined. Using `from __future__ import annotations` to postpone type checking or creating a generic variable: `T = TypeVar('T', bound='_StateFile')` didn't seem to be working either.

File: src/toil/fileStores/abstractFileStore.py [hunk 7]
```diff
@@ -430,14 +432,14 @@ class _StateFile(object):
...
-        def __init__(self, stateDict):
+        def __init__(self, stateDict: Dict[str, Any]):
             assert isinstance(stateDict, dict)
             self.__dict__.update(stateDict)
 
         @classmethod
         @abstractmethod
         @contextmanager
-        def open(cls, outer=None):
+        def open(cls, outer: Optional[Any] = None) -> Iterator[Any]:
```
Comment (jonathanxu18) [inline src/toil/fileStores/abstractFileStore.py:442]: I was thinking of using `Any` here since outer can be an instance of any calling class and the value returned by the Iterator would depend on that calling class.

File: src/toil/fileStores/__init__.py [hunk 8]
```diff
@@ -23,29 +25,29 @@ class FileID(str):
     Calls into the file store can use bare strings; size will be queried from the job store if unavailable in the ID.
     """
 
-    def __new__(cls, fileStoreID, *args):
-        return super(FileID, cls).__new__(cls, fileStoreID)
+    def __new__(cls, fileStoreID: str, *args: Any) -> 'FileID':
+        return super(FileID, cls).__new__(cls, fileStoreID) # type: ignore
```
Comment (mr-c) [inline src/toil/fileStores/__init__.py:29]:
```suggestion
        return super(FileID, cls).__new__(cls, fileStoreID)
```

File: src/toil/fileStores/nonCachingFileStore.py [hunk 9]
```diff
@@ -70,17 +70,18 @@ def open(self, job: Job) -> Generator[None, None, None]:
...
-    def writeGlobalFile(self, localFileName, cleanup=False):
+    def writeGlobalFile(self, localFileName: str, cleanup: bool=False) -> FileID:
         absLocalFileName = self._resolveAbsoluteLocalPath(localFileName)
         creatorID = self.jobDesc.jobStoreID
         fileStoreID = self.jobStore.writeFile(absLocalFileName, creatorID, cleanup)
         if absLocalFileName.startswith(self.localTempDir):
             # Only files in the appropriate directory should become local files
             # we can delete with deleteLocalFile
-            self.localFileMap[fileStoreID].append(absLocalFileName)
+            self.localFileMap[fileStoreID] = absLocalFileName
```
Comment (adamnovak) [inline src/toil/fileStores/nonCachingFileStore.py:80]: This should not be changed. `localFileMap` can have multiple local files (in a list) for each file store ID. Changing this and `readGlobalFile` in this way I think are why the CI tests are failing with complaints that you are trying to delete `/`. The string you put in gets treated as a collection of (1-character) strings, one of which is `/`.

File: src/toil/fileStores/nonCachingFileStore.py [hunk 10]
```diff
@@ -89,20 +90,20 @@ def readGlobalFile(self, fileStoreID, userPath=None, cache=True, mutable=False,
             localFilePath = self.getLocalTempFileName()
 
         self.jobStore.readFile(fileStoreID, localFilePath, symlink=symlink)
-        self.localFileMap[fileStoreID].append(localFilePath)
+        self.localFileMap[fileStoreID] = localFilePath
```
Comment (adamnovak) [inline src/toil/fileStores/nonCachingFileStore.py:93]: Here's the other instance where the append is really needed.

File: src/toil/fileStores/nonCachingFileStore.py [hunk 11]
```diff
@@ -233,10 +234,10 @@ def _getAllJobStates(workflowDir):
                     raise
 
     @staticmethod
-    def _readJobState(jobStateFileName):
+    def _readJobState(jobStateFileName: str) -> Dict[str, str]:
         with open(jobStateFileName, 'rb') as fH:
             state = dill.load(fH)
-        return state
+        return dict(state)
```
Comment (adamnovak) [inline src/toil/fileStores/nonCachingFileStore.py:240]: Maybe this should be a cast? This is going to make a useless copy of the loaded dict.

**Review by adamnovak — CHANGES_REQUESTED** [review]
It looks like you started replacing some internal NonCachingFileStore state instead of appending to it, which I think is what broke the tests.

File: src/toil/fileStores/nonCachingFileStore.py (hunk 9 above)
Comment (jonathanxu18) [reply src/toil/fileStores/nonCachingFileStore.py]: Got it, just switched it back. I think I just got a bit mixed up from the typing for `localFileMap`.

**Review by mr-c — APPROVED** [review]

**Review by adamnovak — APPROVED** [review]
Looks fixed.
